File: Mediator.py
"""
Mediator
Use a mediator object to encapsulate a series of object interactions.
"""

from abc import ABC, abstractmethod


class Mediator(ABC):
    @abstractmethod
    def broadcast(self, msg):
        pass

class ConcreteMediator(Mediator):
    def broadcast(self, msg):
        print(msg)

class Colleague(ABC):
    def __init__(self, mediator):
        self.mediator = mediator
    
    @abstractmethod
    def send(self, msg):
        pass

class ConcreteColleague(Colleague):
    def __init__(self, mediator):
        super().__init__(mediator)
    
    def send(self, msg):
        self.mediator: Mediator
        self.mediator.broadcast(msg)


if __name__ == "__main__":
    m = ConcreteMediator()
    c = ConcreteColleague(m)
    
    c.send("Hello!")

File: Mediator.cpp
/*****
 * Mediator
 * Use a mediator object to encapsulate a series of object interactions.
*****/

#include <iostream>

using namespace std;


class Mediator { public: virtual void broadcast(const string& msg) = 0; };

class ConcreteMediator : public Mediator { public: void broadcast(const string& msg) override { cout << msg << endl; } };

class Colleague {
    protected:
        Mediator* mediator;
    
    public:
        explicit Colleague(Mediator* m) : mediator(m) {}
        virtual void send(const string& msg) = 0;
};

class ConcreteColleague : public Colleague {
    public:
        using Colleague::Colleague;
        void send(const string& msg) override { mediator->broadcast(msg); };
};


int main() {
    ConcreteMediator m;
    ConcreteColleague c(&m);

    c.send("Hello!");

    return 0;
}
